src/validation.cpp

```diff
@@ -2584,13 +2584,14 @@ bool CheckColdStakeFreeOutput(const CTransaction& tx, const int nHeight)
         // if mnsync is incomplete, we cannot verify if this is a budget block.
         // so we check that the staker is not transferring value to the free output
         if (!masternodeSync.IsSynced()) {
-            // First try finding the previous transaction in database
-            CTransactionRef txPrev; uint256 hashBlock;
-            if (!GetTransaction(tx.vin[0].prevout.hash, txPrev, hashBlock, true))
-                return error("%s : read txPrev failed: %s",  __func__, tx.vin[0].prevout.hash.GetHex());
-            CAmount amtIn = txPrev->vout[tx.vin[0].prevout.n].nValue + GetBlockValue(nHeight);
+            // First try finding the previous transaction in cache/database
+            Coin coinstake;
+            if (!pcoinsTip->GetCoin(tx->vin[0].prevout, coinstake) || coinstake.IsSpent()) {
```

Comment:
nit: `AccessCoin` is slightly more efficient than `GetCoin` (and returns a dummy pruned coin if not found).